### PR TITLE
FIX: indexed pandas bar

### LINKED ISSUE #15162: axes.bar fails when x is int-indexed pandas.Series
#### Bug summary
The following codes fail in matplotlib 3.1.1, while work properly in 3.0.3

#### case 1
**Code for reproduction**
```python
import pandas as pd
from matplotlib import pyplot as plt

df = pd.DataFrame({"x":[1,2,3],"width":[.2,.4,.6]},index=[1,2,3])

plt.figure()
plt.bar(df.x, 1, width=df.width)
plt.show()
```

**Actual outcome**
```
Traceback (most recent call last):
  File "/Desktop/fail_example.py", line 7, in <module>
    plt.bar(df.x, 1, width=df.width)
  File "/.pyenv/versions/3.7.3/lib/python3.7/site-packages/matplotlib/pyplot.py", line 2440, in bar
    **({"data": data} if data is not None else {}), **kwargs)
  File "/.pyenv/versions/3.7.3/lib/python3.7/site-packages/matplotlib/__init__.py", line 1601, in inner
    return func(ax, *map(sanitize_sequence, args), **kwargs)
  File "/.pyenv/versions/3.7.3/lib/python3.7/site-packages/matplotlib/axes/_axes.py", line 2430, in bar
    label='_nolegend_',
  File "/.pyenv/versions/3.7.3/lib/python3.7/site-packages/matplotlib/patches.py", line 707, in __init__
    Patch.__init__(self, **kwargs)
  File "/.pyenv/versions/3.7.3/lib/python3.7/site-packages/matplotlib/patches.py", line 89, in __init__
    self.set_linewidth(linewidth)
  File "/.pyenv/versions/3.7.3/lib/python3.7/site-packages/matplotlib/patches.py", line 368, in set_linewidth
    self._linewidth = float(w)
```

**Expected outcome**
Outcome in 3.0.3:
Figure_1

#### case 2
**Code for reproduction**
```python
import pandas as pd
from matplotlib import pyplot as plt

df = pd.DataFrame({"x":[1,3,10]},index=[1000,2000,3000])

plt.figure()
plt.bar(df.x, 1, width=.2)
plt.show()
```

**Actual outcome**
Figure_2

**Expected outcome**
Outcome in 3.0.3:
Figure_3

**Matplotlib version**

  * Operating system: MacOSX
  * Matplotlib version: 3.1.1 (installed from pip)
  * Matplotlib backend (`print(matplotlib.get_backend())`): MacOSX
  * Python version: 3.7.3
  * Jupyter version (if applicable):
  * Other libraries: pandas 0.24.2

**Cause of the bug**
```
_axes.py, line 2363 in bar:
            x0 = x
            x = np.asarray(self.convert_xunits(x))
            width = self._convert_dx(width, x0, x, self.convert_xunits)
_axes.py, line 2166 in _convert_dx: 
            try:
                x0 = x0[0]
            except (TypeError, IndexError, KeyError):
                x0 = x0
```
I guess x0 is expected to be a scalar after this line, However when x0 is an int-indexed (and 0 is not in its indices) pandas.Series, it is continued to be pandas.Series and causes chained disorders to end in the error above.
To evidence, they worked properly when rewritten as (ugly!):
```
            try:
                x0 = x0[0]
            except (TypeError, IndexError, KeyError):
                try:
                    x0 = x0.iat[0]
                except:
                    x0 = x0
```

### PR DESCRIPTION
**Pull request by jklymak**
## PR Summary

Closes #15162

basically if you do `df = pd.DataFrame({"x":[1,2,3],"width":[.2,.4,.6]},index=[1,2,3])`, then `df.x[0]` errors because that index doesn't exist, which breaks `bar`.

While I appreciate this is a bit *more* code than trying to call `iat(0)`, which is what we did before, this solution doesn't presuppose `iat` exists.  

```python
import pandas as pd
from matplotlib import pyplot as plt
import numpy as np

df = pd.DataFrame({"x":[1.,2.,3.],"width":[.2,.4,.6]}, index=[1, 2, 3])
plt.figure()
plt.bar(df.x, 1., width=df.width)
plt.savefig('/Users/jklymak/downloads/Test1.png')

df = pd.DataFrame({"x":[1,3,10]},index=[1000,2000,3000])

plt.figure()
plt.bar(df.x, 1, width=.2)
plt.savefig('/Users/jklymak/downloads/Test2.png')
```

Now yields the expected (and pre 3.1.0):

Test1
Test2

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by ImportanceOfBeingErnest** [comment]
Is this function used anywhere else then with bars? Should bar also support dataframes, like
 
    df = pd.DataFrame({0:[1,2,3]},index=[1,2,3])

This works with `plt.scatter(df, [3,4,5])`, but probably not with `plt.bar(df, [3,4,5])`, right?

**Comment by jklymak** [comment]
I don't think
```python
df = pd.DataFrame({'boo':[1,2,3]},index=[1,2,3])
plt.figure()
plt.bar(df, 1, width=.2)
```
ever worked (`TypeError: float() argument must be a string or a number, not 'NoneType'` on 3.0.0).  The error with this PR is a bit more mysterious, but that's a relatively perverse data frame (I hope).

**Comment by ImportanceOfBeingErnest** [comment]
Ok, just take `df = pd.DataFrame([1,2,3])` then. I think it would be good to define whether single-column dataframes are supposed to work or not. They do for scatter/plot, so one might expect that, but if they don't, maybe a message that is a bit more helpful than "Your linewidth is wrong" would be useful.

**Comment by jklymak** [comment]
This PR is trying to fix something I broke, not expand the duck typing for bar.

File: lib/matplotlib/axes/_axes.py [hunk 1]
```diff
@@ -2138,33 +2138,14 @@ def _convert_dx(dx, x0, xconv, convert):
             # removes the units from unit packages like `pint` that
             # wrap numpy arrays.
             try:
-                x0 = x0[0]
+                x0 = cbook.safe_first_element(x0)
```
Comment (jklymak) [inline lib/matplotlib/axes/_axes.py]: This is great!  Should have checked book....

**Comment by tacaswell** [comment]
backported it #15172

**Pull request merged**